hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleHederaState.java

```diff
@@ -0,0 +1,505 @@
+        }
+
+        if (!label.equals(StateUtils.computeLabel(md.serviceName(), def.stateKey()))) {
+            throw new IllegalArgumentException(
+                    "A label must be computed based on the same "
+                            + "service name and state key in the metadata!");
+        }
+
+        // Put this metadata into the map
+        final var stateMetadata = services.computeIfAbsent(md.serviceName(), k -> new HashMap<>());
+        stateMetadata.put(def.stateKey(), md);
+
+        // Look for a node, and if we don't find it, then insert the one we were given
+        // If there is not a node there, then set it. I don't want to overwrite the existing node,
+        // because it may have been loaded from state on disk, and the node provided here in this
+        // call is always for genesis. So we may just ignore it.
+        if (findNodeIndex(md.serviceName(), def.stateKey()) == -1) {
+            setChild(getNumberOfChildren(), node);
+        }
+    }
+
+    /**
+     * Removes the node and metadata from the state merkle tree.
+     *
+     * @param serviceName The service name. Cannot be null.
+     * @param stateKey The state key
+     */
+    public void removeServiceState(
+            @NonNull final String serviceName, @NonNull final String stateKey) {
+
+        throwIfImmutable();
+        Objects.requireNonNull(serviceName);
+        Objects.requireNonNull(stateKey);
+
+        // Remove the metadata entry
+        final var stateMetadata = services.get(serviceName);
+        if (stateMetadata != null) {
+            stateMetadata.remove(stateKey);
+        }
+
+        // Remove the node
+        final var index = findNodeIndex(serviceName, stateKey);
+        if (index != -1) {
+            setChild(index, null);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public MerkleHederaState copy() {
+        throwIfImmutable();
+        throwIfDestroyed();
+        setImmutable(true);
+        return new MerkleHederaState(this);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public AddressBook getAddressBookCopy() {
+        // To be implemented by Issue #4200
+        throw new RuntimeException("Not yet implemented");
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public void handleConsensusRound(
+            @NonNull final Round round, @NonNull final SwirldDualState swirldDualState) {
+        if (onHandleConsensusRound != null) {
+            onHandleConsensusRound.accept(round, swirldDualState);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public void preHandle(Event event) {
+        if (onPreHandle != null) {
+            onPreHandle.accept(event);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public MerkleNode migrate(int version) {
+        if (onMigrate != null) {
+            onMigrate.accept(this, version);
```

Comment:
Note that this version is the version of this particular merkle node, which is not the same as the application version that gets bumped each release. Not saying that this is a problem, just want you to be aware.